torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +176,217 @@ def magic_impl(self, other):
+
+    def create_symint(self, name, val):
         if not HAS_SYMPY:
             raise RuntimeError("Need sympy installed to create symbolic shapes")
-        if shape_env is None:
-            shape_env = self.shape_env
+
         # Currently we don't put 0/1 specialization in guards but perhaps we should
         if val == 0 or val == 1:
             return val
         sympy_expr = sympy.Symbol(name, positive=True, integer=True)
         py_sym_int = PySymInt(sympy_expr, self)
         cpp_sym_int = torch.SymIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
-        shape_env[sympy_expr] = val
+        self.var_to_val[sympy_expr] = sympy.Integer(val)
         return cpp_sym_int
 
-    def try_constantify(self, expr):
-        # Simplifies assuming that shape vars > 1 (since we cache on 0/1 shape values)
-        new_shape_env = {
-            k: sympy.Symbol(f"shape_{idx}", positive=True, integer=True) + 1
-            for idx, k in enumerate(self.shape_env.keys())
-        }
-        new_expr = expr.subs(new_shape_env)
-        new_expr = new_expr.simplify()
-        if len(list(new_expr.free_symbols)) == 0:
-            return new_expr
-        return None
-
-    def create_shapes_for_args(self, args, shape_env=None):
-        # Takes pytrees and returns a flat list
+    def create_shapes_for_args(self, args):
         arg_cnt = 0
 
         def create_shape(x):
             nonlocal arg_cnt
             if not isinstance(x, torch.Tensor):
                 return x
 
-            out_shape = [self.create_symint(f"s_{arg_cnt}[{idx}]", sz, shape_env) for idx, sz in enumerate(x.shape)]
+            out_shape = [self.create_symint(f"s_{arg_cnt}[{idx}]", sz) for idx, sz in enumerate(x.shape)]
             arg_cnt += 1
             return out_shape
         return list(map(create_shape, pytree.tree_flatten(args)[0]))
 
     def evaluate_guards_for_args(self, *args):
-        env: Dict[Any, Any] = {}
-        _ = self.create_shapes_for_args(args, shape_env=env)
-        return all(guard.subs(env) == value for guard, value in self.guards)
-
-
-    def evaluate_expr(self, expr):
-        const_expr = self.try_constantify(expr)
-        if const_expr is not None:
-            return const_expr
-
-        expr = expr.simplify()
-        concrete_val = expr.subs(self.shape_env)
-
-        # Uncomment this to see what code triggered this guard.
-        # TODO: Save this to the guard representation so you can look
-        # at it later
-        # import traceback
-        # traceback.print_stack()
+        new_env = ShapeEnv()
+        _ = new_env.create_shapes_for_args(args)
+        return all(guard.xreplace(new_env.var_to_val) == value for guard, value, _ in self.guards)
+
+    def get_nontrivial_guards(self):
+        guards = [(self.simplify(guard), val) for guard, val, _ in self.guards]
+        guards = [guard for guard in guards if len(guard[0].free_symbols) > 0]
+        return guards
```

Comment:
Should this update guards field so we don't have to resimplify?

Reply:
For now I'm just using this for testing/debugging purposes. Will take a longer think at API when we get to integrating with dynamo.